config/hotreloadable.go

```diff
@@ -143,115 +433,335 @@ func (c *Config) RegisterDurationConfigVariable(defaultValueInTimescaleUnits int
+		*ptr = v
+	}, orderedKeys...)
+}
+
+// RegisterStringMapVar registers a not hot-reloadable string map config variable
+// WARNING: keys are being looked up in requested order and the value of the first found key is returned,
+// e.g. asking for the same keys but in a different order can result in a different value to be returned
+func (c *Config) RegisterStringMapVar(
+	defaultValue map[string]interface{}, orderedKeys ...string,
+) map[string]interface{} {
+	var ret map[string]interface{}
+	c.registerStringMapVar(defaultValue, nil, false, func(v map[string]interface{}) {
+		ret = v
+	}, orderedKeys...)
+	return ret
+}
+
+// RegisterReloadableStringMapVar registers a hot-reloadable string map config variable
+// WARNING: keys are being looked up in requested order and the value of the first found key is returned,
+// e.g. asking for the same keys but in a different order can result in a different value to be returned
+func (c *Config) RegisterReloadableStringMapVar(
+	defaultValue map[string]interface{}, orderedKeys ...string,
+) *Reloadable[map[string]interface{}] {
+	ptr := getOrCreatePointer(
+		c.reloadableVars, c.reloadableVarsMisuses, &c.reloadableVarsLock, defaultValue, orderedKeys...,
+	)
+	c.registerStringMapVar(defaultValue, ptr, true, func(v map[string]interface{}) {
+		ptr.store(v)
+	}, orderedKeys...)
+	return ptr
+}
+
+func (c *Config) registerStringMapVar(
+	defaultValue map[string]interface{}, ptr any, isHotReloadable bool, store func(map[string]interface{}), orderedKeys ...string,
+) {
 	c.vLock.RLock()
 	defer c.vLock.RUnlock()
 	c.hotReloadableConfigLock.Lock()
 	defer c.hotReloadableConfigLock.Unlock()
 	configVar := configValue{
 		value:        ptr,
 		defaultValue: defaultValue,
-		keys:         keys,
+		keys:         orderedKeys,
 	}
 
 	if isHotReloadable {
-		c.appendVarToConfigMaps(keys[0], &configVar)
+		c.appendVarToConfigMaps(orderedKeys, &configVar)
 	}
 
-	for _, key := range keys {
+	for _, key := range orderedKeys {
 		if c.IsSet(key) {
-			*ptr = c.GetStringMap(key, defaultValue)
+			store(c.GetStringMap(key, defaultValue))
 			return
 		}
 	}
-	*ptr = defaultValue
+	store(defaultValue)
 }
 
-func (c *Config) appendVarToConfigMaps(key string, configVar *configValue) {
+func (c *Config) appendVarToConfigMaps(keys []string, configVar *configValue) {
+	key := strings.Join(keys, ",")
 	if _, ok := c.hotReloadableConfig[key]; !ok {
 		c.hotReloadableConfig[key] = make([]*configValue, 0)
 	}
 	c.hotReloadableConfig[key] = append(c.hotReloadableConfig[key], configVar)
 }
+
+type configTypes interface {
+	int | int64 | string | time.Duration | bool | float64 | []string | map[string]interface{}
+}
+
+// Reloadable is used as a wrapper for hot-reloadable config variables
+type Reloadable[T configTypes] struct {
+	value T
+	lock  sync.RWMutex
+}
+
+// Load should be used to read the underlying value without worrying about data races
+func (a *Reloadable[T]) Load() T {
```

Comment:
**Question:** is `Get` or `Read` better than `Load`?

Reply:
No real preference, `Load()` reminds me of the `atomic.Value` way of retrieving the value so it is familiar to me. Leaving this here as unresolved, maybe someone can also review this and express a preference :+1: